Performance improve on overlap_with_displaced_states()

JiakaiW opened the pull request:
When I had very long omega_d_values and very long chi_ac_values, it seems FloquetAnalysis.run() spends most of the time on the overlap_with_displaced_states() calls and I observe CPU utilization to be very low. 
Somehow this new overlap_with_displaced_states is much faster. Mostly because explicit numpy usage.

dkweiss31 commented:
Hi @JiakaiW, thanks for the PR! I'm wondering if you have a MWE where you run into this issue, and maybe also some profiling data? It would be nice to verify that I see the same slowdown on my machine, for instance.

JiakaiW commented:
> Hi @JiakaiW, thanks for the PR! I'm wondering if you have a MWE where you run into this issue, and maybe also some profiling data? It would be nice to verify that I see the same slowdown on my machine, for instance.

Here's an example in a notebook. In this branch I added a "new_functions" attribute to ft.Options, and I added print messages in run(), and progress bars to the parallel map.  

https://github.com/JiakaiW/floquet_fork/blob/testing/notebooks/comparing_speed.ipynb

dkweiss31 commented:
Right, so more or less this MWE?

```python
import numpy as np
import qutip as qt
import scqubits as scq

import floquet as ft

num_states = 20
qubit_params = {"EJ": 20.0, "EC": 0.2, "ng": 0.25, "ncut": 41}
tmon = scq.Transmon(**qubit_params, truncated_dim=num_states)
state_indices = [0, 1]


def get_H0_H1(qubit_instance: scq.GenericQubit) -> tuple[qt.Qobj, qt.Qobj]:
    hilbert_space = scq.HilbertSpace([qubit_instance])
    hilbert_space.generate_lookup()
    evals = hilbert_space["evals"][0][0:num_states]
    H0 = 2.0 * np.pi * qt.Qobj(np.diag(evals - evals[0]))
    H1 = hilbert_space.op_in_dressed_eigenbasis(qubit_instance.n_operator)
    return H0, H1


H0, H1 = get_H0_H1(tmon)
omega_d_values = 2.0 * np.pi * np.linspace(7.5, 10.0, 200)
chi_ac_values = 2.0 * np.pi * np.linspace(0.0, 0.1, 200)
chi_to_amp = ft.ChiacToAmp(H0, H1, state_indices, omega_d_values)
drive_amplitudes = chi_to_amp.amplitudes_for_omega_d(chi_ac_values)

model = ft.Model(
    H0, H1, omega_d_values=omega_d_values, drive_amplitudes=drive_amplitudes
)

options = ft.Options(
    fit_range_fraction=0.5,
    floquet_sampling_time_fraction=0.0,
    fit_cutoff=4,
    overlap_cutoff=0.8,
    nsteps=30_000,
    num_cpus=4,
    save_floquet_modes=True,
)

floquet_analysis = ft.FloquetAnalysis(
    model, state_indices=state_indices, options=options
)
data_vals = floquet_analysis.run()
```
I have tried profiling this example using [py-spy](https://github.com/benfred/py-spy) and [speedscope](https://www.speedscope.app/): if this example is in the file `profile.py`, I call
```
sudo py-spy record -o profile.json --subprocesses --rate 2 -f speedscope -- python profile.py
```
and then upload the result to speedscope. We need the subprocesses flag to correctly profile child processes. Here is a screenshot for one of the threads, which is representative: 
<img width="1390" alt="Screenshot 2025-04-25 at 8 12 17 AM" src="https://github.com/user-attachments/assets/342179d3-fb42-40c3-be76-146290c0eba9" />
you can see I'm hovering my mouse over the `_run_overlap_displaced` call. It seems to occupy a small amount of compute time as compared to `_run_floquet_and_calculate`. So I'm not positive that the overlap calculations are contributing to the low cpu utilizations you're observing. Can you try running this and sharing the result? 

At the same time, below are some possible issues to check for:

1. I see in your notebook you have `num_cpus=20`. Does the machine you're running this on have at least 20 cores? If not, this will slow down your code. You want to set this number to the number of cores available on your machine, or fewer.
2. I've never had luck using progress bars in multiprocessing calculations, though it seems there has been progress in recent years. Is there a noticeable runtime difference if you remove the progress bar?

JiakaiW commented:
You are right! The whole issue is caused by the progress bar. 
I tested three versions: 1) the original, everything unchanged 2) With progress bar, 3) with progress bar and new overlap_with_displaced_states. And I found the new overlap_with_displaced_states happens to preserve the performance of no progress bar. 
```bash
#! /bin/bash

# The purpose is to run the profiling tool with different installed versions.

# Store the base directory path
BASE_DIR="$(cd "$(dirname "$0")/.." && pwd)"
PROFILE_DIR="$BASE_DIR/profile"

# First, run the original version.
cd "$BASE_DIR"
pip uninstall -y floquet
pip install -e .
cd "$PROFILE_DIR"
sudo py-spy record -o origional.json --subprocesses --rate 2 -f speedscope -- python test_floquet.py

# Then, run the tqdm version without new overlap_with_displaced_states.
cd "$BASE_DIR"
pip uninstall -y floquet
cd floquet/utils
mv parallel.py parallel_original.py
cp parallel_modified.py parallel.py
cd "$BASE_DIR"
pip install -e .
cd "$PROFILE_DIR"
sudo py-spy record -o tqdm_with_origional_overlap_with_displaced_states.json --subprocesses --rate 2 -f speedscope -- python test_floquet.py

# Then, run the tqdm version with new overlap_with_displaced_states.
cd "$BASE_DIR"
pip uninstall -y floquet
cd floquet
mv displaced_state.py displaced_state_original.py
cp displaced_state_modified.py displaced_state.py
cd "$BASE_DIR"
pip install -e .
cd "$PROFILE_DIR"
sudo py-spy record -o tqdm_with_new_overlap_with_displaced_states.json --subprocesses --rate 2 -f speedscope -- python test_floquet.py

# Cleanup - restore original files
cd "$BASE_DIR/floquet/utils"
mv parallel_original.py parallel.py
cd "$BASE_DIR/floquet"
mv displaced_state_original.py displaced_state.py
```
my test_floquet.py is 
```python
import numpy as np
import qutip as qt
import scqubits as scq
import floquet as ft


def main():
    num_states = 20
    qubit_params = {"EJ": 20.0, "EC": 0.2, "ng": 0.25, "ncut": 41}
    tmon = scq.Transmon(**qubit_params, truncated_dim=num_states)
    state_indices = [0, 1]


    def get_H0_H1(qubit_instance: scq.GenericQubit) -> tuple[qt.Qobj, qt.Qobj]:
        hilbert_space = scq.HilbertSpace([qubit_instance])
        hilbert_space.generate_lookup()
        evals = hilbert_space["evals"][0][0:num_states]
        H0 = 2.0 * np.pi * qt.Qobj(np.diag(evals - evals[0]))
        H1 = hilbert_space.op_in_dressed_eigenbasis(qubit_instance.n_operator)
        return H0, H1


    H0, H1 = get_H0_H1(tmon)
    omega_d_values = 2.0 * np.pi * np.linspace(7.5, 10.0, 100)
    chi_ac_values = 2.0 * np.pi * np.linspace(0.0, 0.1, 100)
    chi_to_amp = ft.ChiacToAmp(H0, H1, state_indices, omega_d_values)
    drive_amplitudes = chi_to_amp.amplitudes_for_omega_d(chi_ac_values)

    model = ft.Model(
        H0, H1, omega_d_values=omega_d_values, drive_amplitudes=drive_amplitudes
    )

    options = ft.Options(
        fit_range_fraction=0.5,
        floquet_sampling_time_fraction=0.0,
        fit_cutoff=4,
        overlap_cutoff=0.8,
        nsteps=30_000,
        num_cpus=4,
        save_floquet_modes=True,
    )

    floquet_analysis = ft.FloquetAnalysis(
        model, state_indices=state_indices, options=options
    )
    data_vals = floquet_analysis.run()
if __name__ == "__main__":
    main()
```
The original one took 40seconds
<img width="1019" alt="image" src="https://github.com/user-attachments/assets/a838ab9b-bc4f-4b16-90a4-b006a01ab6c5" />
The one with progressbar is plagued by communication overhead between the works and the io:
<img width="1023" alt="image" src="https://github.com/user-attachments/assets/4ffe2666-0c73-4a61-ade6-247bbf3fda32" />

And the one with new overlap_with_displaced_states doesn't suffer from the progress bar
<img width="1019" alt="image" src="https://github.com/user-attachments/assets/60a99d21-9fd7-480e-911d-900873f73dfe" />

JiakaiW commented:
And it seems the new overlap_with_displaced_states is faster than the old one, the speedup most likely comes from batching of jobs, and using explicit numpy operations

dkweiss31 commented:
Nice! It seems indeed like the progress bar was slowing things down. Now the current version is only slightly slower than the version you've proposed. I'd like to really nail down where the slowdown is coming from: that way we can really make targeted changes to this calculation. Is it possible for you to profile this function? What I mean is, I'm not positive that the parallel logic needs to change: what if we replaced the call 
```python
disp_state = self.displaced_state(
    omega_d,
    amp,
    state_idx=state_idx,
    coefficients=coefficients[array_idx],
).dag()
```
with instead the more numpy vectorized call that you have. Does that simple replacement maintain the speedup?
